### PR TITLE
Make initiateKeyTransfer() async

### LINKED ISSUE #107: Make dc.initiateKeyTransfer() async
See https://github.com/deltachat/deltachat-desktop/issues/33 for details

### LINKED ISSUE #124: Make dc.continueKeyTransfer() async
See https://github.com/deltachat/deltachat-desktop/issues/33 for details

### PR DESCRIPTION
**Pull request by Jikstra**
Closes #124, closes #107

### CODE REVIEW

File: test/integration.js [hunk 1]
```diff
@@ -365,6 +365,14 @@ test('ChatList methods', t => {
   t.end()
 })
 
+test('key transfer methods', t => {
+  dc.initiateKeyTransfer((err, setupCode) => {
+    t.same(err, null, 'err is null')
```
Comment (ralphtheninja) [inline test/integration.js]: Actually, we should use the same thing for errors, e.g. `t.error(err, 'no error')`, see https://github.com/substack/tape#terrorerr-msg

Comment (ralphtheninja) [reply test/integration.js]: I should have said this from the start, sorry about that.

Comment (Jikstra) [reply test/integration.js]: > I should have said this from the start, sorry about that.

no problem ^^

File: index.js [hunk 2]
```diff
@@ -155,14 +155,13 @@ class DeltaChat extends EventEmitter {
...
-  continueKeyTransfer (messageId, setupCode) {
-    return Boolean(
-      binding.dcn_continue_key_transfer(
-        this.dcn_context,
-        Number(messageId),
-        setupCode
-      )
-    )
+  continueKeyTransfer (messageId, setupCode, cb) {
+    binding.dcn_continue_key_transfer(this.dcn_context, Number(messageId), setupCode, result => {
```
Comment (ralphtheninja) [inline index.js]: :+1:

File: src/napi-macros-extensions.h [hunk 3]
```diff
@@ -41,3 +41,47 @@
...
+  NAPI_STATUS_THROWS(napi_get_reference_value(env, carrier->callback_ref, &callback)); \
+  NAPI_STATUS_THROWS(napi_call_function(env, global, callback, argc, argv, NULL)); \
+  NAPI_STATUS_THROWS(napi_delete_reference(env, carrier->callback_ref)); \
+  NAPI_STATUS_THROWS(napi_delete_async_work(env, carrier->async_work)); \
+
+#define NAPI_ASYNC_NEW_CARRIER(name) \
+  name##_carrier_t* carrier = calloc(1, sizeof(name##_carrier_t)); \
+  carrier->dcn_context = dcn_context; \
+
+#define NAPI_ASYNC_QUEUE_ASYNC_WORK_AND_RETURN(name, cb) \
```
Comment (ralphtheninja) [inline src/napi-macros-extensions.h]: Maybe call it `NAPI_ASYNC_QUEUE_WORK()` and make `NAPI_RETURN_UNDEFINED()` explicit? Functions might need to return something, even though they do async work.

Comment (ralphtheninja) [inline src/napi-macros-extensions.h]: Maybe call it `NAPI_ASYNC_QUEUE_WORK()` and make `NAPI_RETURN_UNDEFINED()` explicit? Functions might need to return something, even though they do async work. Also the name is painfully long and suggests it should be broken down :smile:

File: src/test.c [hunk 4]
```diff
@@ -0,0 +1,15 @@
...
+
+#define NAPI_ASYNC_CARRIER_END(name) \
+  } name##_carriert_t; \
+
+
+NAPI_ASYNC_CARRIER_BEGIN(test)
+  int msg_id;
+  char* setup_code;
+  int result;
+NAPI_ASYNC_CARRIER_END(test)
```
Comment (ralphtheninja) [inline src/test.c]: What does this file do?

File: src/napi-macros-extensions.h [hunk 5]
```diff
@@ -41,3 +41,47 @@
...
+  napi_value callback = cb; \
+  napi_value async_resource_name; \
+  NAPI_STATUS_THROWS(napi_create_reference(env, callback, 1, &carrier->callback_ref)); \
+  NAPI_STATUS_THROWS(napi_create_string_utf8(env, "name##_callback", \
```
Comment (ralphtheninja) [inline src/napi-macros-extensions.h]: `"name##_callback"` is a string that will be identical for _all_ resource names. You probably want to do something like `#name ## "_callback"`. See examples in `napi-macros` https://github.com/mafintosh/napi-macros/blob/master/napi-macros.h#L22

File: test/integration.js [hunk 6]
```diff
@@ -7,6 +7,24 @@ const c = require('../constants')
 
 let dc = null
 
+function configureDefaultDC (dcInstance) {
```
Comment (ralphtheninja) [inline test/integration.js]: Can call this parameter `dc`.

File: test/integration.js [hunk 7]
```diff
@@ -365,7 +369,32 @@ test('ChatList methods', t => {
   t.end()
 })
 
-test('key transfer methods', t => {
+test.skip('key transfer', t => {
```
Comment (ralphtheninja) [inline test/integration.js]: In a test like this you should use `t.plan(N)` where `N` is the number of tests you expect (and remove `t.end()`).

**Comment by ralphtheninja** [comment]
> @ralphtheninja it's hard to implement tests for continueKeyTransfer() without initiateKeyTransfer(), therefore i put it into this pr.

Good call!

File: index.js [hunk 8]
```diff
@@ -155,14 +155,13 @@ class DeltaChat extends EventEmitter {
...
+      if (result === 0) {
+        return cb(new Error('Key transfer failed due bad setup code'))
```
Comment (ralphtheninja) [inline index.js]: `due to`

File: index.js [hunk 9]
```diff
@@ -362,8 +361,13 @@ class DeltaChat extends EventEmitter {
...
   }
 
-  initiateKeyTransfer () {
-    return binding.dcn_initiate_key_transfer(this.dcn_context)
+  initiateKeyTransfer (cb) {
+    return binding.dcn_initiate_key_transfer(this.dcn_context, statusCode => {
+      if (typeof statusCode === 'string') {
+        return cb(null, statusCode)
+      }
+      cb(new Error('could not initiate key transfer'))
```
Comment (ralphtheninja) [inline index.js]: `Could not..`

File: test/integration.js [hunk 10]
```diff
@@ -365,6 +369,39 @@ test('ChatList methods', t => {
...
+  })
+
+  dc2.once('ready', () => {
+    dc.initiateKeyTransfer((err, setupCode) => {
+      t.error(err, 'no err')
+      t.is(typeof setupCode, 'string', 'setupCode is string')
+    })
+  })
+
+  dc2.open(cwd2, err => {
```
Comment (ralphtheninja) [inline test/integration.js]: Can move `tempy.directory()` directly here and remove `cwd2`.

**Comment by ralphtheninja** [comment]
@Jikstra A tip. When creating a PR that solves some issues, you can do write `Closes #issue1, closes #issue2` etc and once the PR is merged the corresponding issues will be automatically closed.

**Comment by ralphtheninja** [comment]
Oh, and the README needs to be updated as well, since the signature has been changed for these two methods.

File: src/napi-macros-extensions.h (hunk 3 above)
Comment (Jikstra) [reply src/napi-macros-extensions.h]: yup, you're right

File: src/test.c (hunk 4 above)
Comment (Jikstra) [reply src/test.c]: oh this was just a test file to debug the preprocessor stuff (gcc -E ftw). Will git rm.

File: src/napi-macros-extensions.h (hunk 5 above)
Comment (Jikstra) [reply src/napi-macros-extensions.h]: not `#name "callback"` ?

File: test/integration.js (hunk 6 above)
Comment (Jikstra) [reply test/integration.js]: Can confuse with global dc but yes.

File: test/integration.js (hunk 7 above)
Comment (Jikstra) [reply test/integration.js]: Didn't understand the plan method. What kind of iterations? Does it run the test multiple times or what's the reason of it?

File: src/test.c (hunk 4 above)
Comment (ralphtheninja) [reply src/test.c]: Aaah cool. Very handy indeed! Can use it to debug the resource names :wink:

File: test/integration.js (hunk 6 above)
Comment (ralphtheninja) [reply test/integration.js]: No, it can't confuse :) It's a single parameter in a one line function.

File: README.md [hunk 11]
```diff
@@ -283,10 +287,12 @@ Import/export things. Corresponds to [`dc_imex()`](https://deltachat.github.io/a
 
 Check if there is a backup file. Corresponds to [`dc_imex_has_backup()`](https://deltachat.github.io/api/classdc__context__t.html#a052b3b20666162d35b57b34cecf74888).
 
-#### `dc.initiateKeyTransfer()`
+#### `dc.initiateKeyTransfer([callback])`
```
Comment (ralphtheninja) [inline README.md]: Callback is _not_ optional.

File: README.md [hunk 12]
```diff
@@ -283,10 +287,12 @@ Import/export things. Corresponds to [`dc_imex()`](https://deltachat.github.io/a
...
 
 Initiate Autocrypt setup transfer. Corresponds to [`dc_initiate_key_transfer()`](https://deltachat.github.io/api/classdc__context__t.html#af327aa51e2e18ce3f5948545a637eac9).
 
+* `callback` *(function, optional)* Called with an error as first argument (or null) and the setup code as second argument if no error occured.
```
Comment (ralphtheninja) [inline README.md]: Not optional.

File: README.md [hunk 13]
```diff
@@ -135,10 +135,14 @@ The `options` object takes the following properties:
...
 * `options.e2ee_enabled` *(boolean, optional)*: Enable E2EE. Defaults to `true`.
 
-#### `dc.continueKeyTransfer(messageId, setupCode)`
+#### `dc.continueKeyTransfer(messageId, setupCode, callback)`
 
 Continue the AutoCrypt key transfer on another device. Corresponds to [`dc_continue_key_transfer()`](https://deltachat.github.io/api/classdc__context__t.html#a5af2cdd80c7286b2a495d56fa6c0832f).
 
+* `messageId` *(int, required)* See deltachat api documentation
+* `setupCode` *(string, required)* See deltachat api documentation
+* `callback` *(function, optional)* Called with an error if setup code is bad
```
Comment (ralphtheninja) [inline README.md]: The callback is not optional

File: README.md [hunk 14]
```diff
@@ -135,10 +135,14 @@ The `options` object takes the following properties:
 * `options.selfstatus` *(string, optional)*: Own status to display e.g. in email footers, defaults to a standard text.
 * `options.e2ee_enabled` *(boolean, optional)*: Enable E2EE. Defaults to `true`.
 
-#### `dc.continueKeyTransfer(messageId, setupCode)`
+#### `dc.continueKeyTransfer(messageId, setupCode, callback)`
 
 Continue the AutoCrypt key transfer on another device. Corresponds to [`dc_continue_key_transfer()`](https://deltachat.github.io/api/classdc__context__t.html#a5af2cdd80c7286b2a495d56fa6c0832f).
 
+* `messageId` *(int, required)* See deltachat api documentation
```
Comment (ralphtheninja) [inline README.md]: string|integer

File: README.md (hunk 11 above)
Comment (Jikstra) [reply README.md]: Callback on dc.open is also not optional than?

Comment (ralphtheninja) [reply README.md]: Correct! :smile:

Comment (ralphtheninja) [reply README.md]: I can fix that in another commit.

File: README.md (hunk 14 above)
Comment (Jikstra) [reply README.md]: are you sure about the string?

Comment (ralphtheninja) [reply README.md]: Yes, since you do `Number(messageId)` in the code.

Comment (Jikstra) [reply README.md]: oh correct, only looked at the core docu.

**Comment by Jikstra** [comment]
Will squash as soon you're happy @ralphtheninja :)

File: README.md (hunk 11 above)
Comment (ralphtheninja) [reply README.md]: Brackets should be removed.

File: test/integration.js (hunk 7 above)
Comment (Jikstra) [reply test/integration.js]: would like to skip, still don't really get it :D

File: test/integration.js (hunk 6 above)
Comment (ralphtheninja) [reply test/integration.js]: `dc`, it's not _super_ important

**Comment by ralphtheninja** [comment]
> Will squash as soon you're happy @ralphtheninja :)

Please squash here on github instead of doing funky `git rebase -i HEAD~whatever` :wink:

**Comment by ralphtheninja** [comment]
:shipit:

File: test/integration.js (hunk 7 above)
Comment (ralphtheninja) [reply test/integration.js]: The idea is to ensure all sub tests are run before the whole test ends. In theory, this test _could_ end when `t.end()` is called, even though some of the other sub tests haven't been executed.

**Pull request merged**